base/io.jl

```diff
@@ -669,8 +669,34 @@ function readuntil_indexable(io::IO, target#=::Indexable{T}=#, out)
     if done(target, first)
         return true
     end
-    len = endof(target)
+
+    # In readuntil, we use a cache to accelerate scanning of sequences
+    # with repeated patterns. Each cache entry tells us which index
+    # we should start the search at. When we allocate the cache,
+    # we initialize it to 0 (and offset by the first index afterwards).
+    # Suppose target is:
+    #    Index:  1245689
+    #    Value: "aδcaδcx"
+    # We would set the cache to
+    #    0 0 0 1 0 3 4 0
+    # So after if we mismatch after the second aδc sequence,
+    # we can pick back up at index 5 (4 + 1).
+    function grow_readuntil_cache!(cache, target, max_pos, pos)
```

Comment:
It still feels that it's weird to separate this into a different function. It happens to currently be true that the call can be defined in simple terms of `(cache, target, max_pos, pos) => max_pos`, but if you look back at the history of making this change, that's simply an artifact of our choice of assumptions about how to operate efficient on our current String or Array API in typical cases.